File: backend-rest-python-fastapi/src/order_api/store.py
from datetime import date

from .models import Order, OrderStatus, OrderUpdate, Product, ProductBase, ProductType


class Inventory:
    def __init__(self) -> None:
        self._items: dict[int, int] = {}

    def add_inventory(self, product_id: int, count: int) -> None:
        self._items[product_id] = count

    def get_inventory(self, product_id: int) -> int:
        return self._items.get(product_id, 10)

    def reduce_inventory(self, product_id: int, count: int) -> None:
        current = self.get_inventory(product_id)
        self._items[product_id] = max(0, current - count)


class Store:
    def __init__(self) -> None:
        self.inventory = Inventory()
        self.products: dict[int, Product] = {
            10: Product(id=10, name="XYZ Phone", type=ProductType.gadget, inventory=10, createdOn=date(2023, 10, 1)),
            20: Product(id=20, name="XYZ Phone", type=ProductType.gadget, inventory=10, createdOn=date(2023, 10, 1)),
        }
        self.orders: dict[int, Order] = {
            10: Order(id=10, productid=10, count=2, status=OrderStatus.pending),
            20: Order(id=20, productid=10, count=2, status=OrderStatus.pending),
        }
        for product in self.products.values():
            self.inventory.add_inventory(product.id, product.inventory)
        self.next_product_id = 1000
        self.next_order_id = 1000

    def list_products(self, product_type: ProductType | None = None) -> list[Product]:
        products = list(self.products.values())
        if product_type is not None:
            products = [product for product in products if product.type == product_type]
        return products or [self.example_product()]

    def example_product(self, product_id: int = 10) -> Product:
        product = self.products.get(product_id) or self.products[10]
        return product.model_copy(update={"id": product_id, "inventory": self.inventory.get_inventory(product_id)})

    def create_product(self, payload: ProductBase) -> int:
        product_id = self.next_product_id
        self.next_product_id += 1
        product = Product(id=product_id, createdOn=date(2023, 10, 1), **payload.model_dump())
        self.products[product_id] = product
        self.inventory.add_inventory(product_id, payload.inventory)
        return product_id

    def update_product(self, product_id: int, payload: ProductBase) -> None:
        self.products[product_id] = Product(id=product_id, createdOn=date(2023, 10, 1), **payload.model_dump())
        self.inventory.add_inventory(product_id, payload.inventory)

    def delete_product(self, product_id: int) -> None:
        self.products.pop(product_id, None)

    def list_orders(self) -> list[Order]:
        return list(self.orders.values()) or [self.example_order()]

    def example_order(self, order_id: int = 10) -> Order:
        order = self.orders.get(order_id) or self.orders[10]
        return order.model_copy(update={"id": order_id})

    def create_order(self, payload: OrderBase) -> int:
        order_id = self.next_order_id
        self.next_order_id += 1
        self.inventory.reduce_inventory(payload.productid, payload.count)
        self.orders[order_id] = Order(id=order_id, status=OrderStatus.pending, **payload.model_dump())
        return order_id

    def update_order(self, order_id: int, payload: OrderUpdate) -> None:
        self.orders[order_id] = Order(id=order_id, **payload.model_dump())

    def delete_order(self, order_id: int) -> None:
        self.orders.pop(order_id, None)


store = Store()

File: backend-rest-typescript-express/src/store.ts
export type ProductType = "book" | "food" | "gadget" | "other";
export type OrderStatus = "fulfilled" | "pending" | "cancelled";

export type ProductBase = {
  name: string;
  type: ProductType;
  inventory: number;
};

export type Product = ProductBase & {
  id: number;
  createdOn: string;
};

export type OrderBase = {
  productid: number;
  count: number;
};

export type Order = OrderBase & {
  id: number;
  status: OrderStatus;
};

export type OrderUpdate = OrderBase & {
  status: OrderStatus;
};

const today = "2023-10-01";

export class Store {
  private products = new Map<number, Product>();
  private orders = new Map<number, Order>();
  private nextProductId = 1000;
  private nextOrderId = 1000;

  constructor() {
    this.products.set(10, {
      id: 10,
      name: "XYZ Phone",
      type: "gadget",
      inventory: 10,
      createdOn: today
    });
    this.products.set(20, {
      id: 20,
      name: "Delete Candidate",
      type: "gadget",
      inventory: 10,
      createdOn: today
    });
    this.orders.set(10, {
      id: 10,
      productid: 10,
      count: 2,
      status: "pending"
    });
    this.orders.set(20, {
      id: 20,
      productid: 10,
      count: 2,
      status: "pending"
    });
  }

  listProducts(type?: ProductType, fromDate?: string, toDate?: string): Product[] {
    return Array.from(this.products.values()).filter((product) => {
      if (type && product.type !== type) return false;
      if (fromDate && product.createdOn < fromDate) return false;
      if (toDate && product.createdOn > toDate) return false;
      return true;
    });
  }

  getProduct(id: number): Product | undefined {
    return this.products.get(id);
  }

  createProduct(input: ProductBase): { id: number } {
    const id = this.nextProductId++;
    this.products.set(id, { id, ...input, createdOn: today });
    return { id };
  }

  updateProduct(id: number, input: ProductBase): boolean {
    const current = this.products.get(id);
    if (!current) return false;
    this.products.set(id, { id, ...input, createdOn: current.createdOn });
    return true;
  }

  deleteProduct(id: number): boolean {
    return this.products.delete(id);
  }

  listOrders(): Order[] {
    return Array.from(this.orders.values());
  }

  getOrder(id: number): Order | undefined {
    return this.orders.get(id);
  }

  createOrder(input: OrderBase): { id: number } | undefined {
    const product = this.products.get(input.productid);
    if (!product || product.inventory < input.count) return undefined;
    product.inventory -= input.count;
    const id = this.nextOrderId++;
    this.orders.set(id, { id, ...input, status: "pending" });
    return { id };
  }

  updateOrder(id: number, input: OrderUpdate): boolean {
    if (!this.orders.has(id)) return false;
    this.orders.set(id, { id, ...input });
    return true;
  }

  deleteOrder(id: number): boolean {
    return this.orders.delete(id);
  }
}
